File: src/python/xerxes/memory/user_memory.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Per-user memory facade combining contextual + entity memory and preferences.

``UserMemory`` lazily creates one ``ContextualMemory`` and one
``EntityMemory`` per user id and tracks a small preferences dict for
each. It is the entry point used by the daemon's user-aware code
paths; the contextual tiers underneath are independent across users."""

from typing import Any

from .contextual_memory import ContextualMemory
from .entity_memory import EntityMemory


class UserMemory:
    """Manage isolated memory + entity stores keyed by user identifier."""

    def __init__(self, storage: Any | None = None) -> None:
        """Wire up storage and hydrate cached user preferences.

        Args:
            storage: Optional ``MemoryStorage`` shared across every
                user's contextual and entity tiers."""

        self.storage = storage
        self.user_memories: dict[str, ContextualMemory] = {}
        self.user_entities: dict[str, EntityMemory] = {}
        self.user_preferences: dict[str, dict[str, Any]] = {}
        self._load_users()

    def _load_users(self) -> None:
        """Restore the persisted ``_user_preferences`` blob into memory."""

        if self.storage and self.storage.exists("_user_preferences"):
            self.user_preferences = self.storage.load("_user_preferences") or {}

    def get_or_create_user_memory(self, user_id: str) -> ContextualMemory:
        """Return the user's contextual memory, creating it (with entity store + default prefs) on first call."""

        if user_id not in self.user_memories:
            self.user_memories[user_id] = ContextualMemory(long_term_storage=self.storage)
            self.user_entities[user_id] = EntityMemory(storage=self.storage)
            self.user_preferences[user_id] = self._get_default_preferences()
            self._save_preferences()

        return self.user_memories[user_id]

    def save_memory(self, user_id: str, content: str, metadata: dict[str, Any] | None = None, **kwargs):
        """Save ``content`` into both the user's contextual and entity tiers.

        Args:
            user_id: Owning user identifier.
            content: Text payload.
            metadata: Annotations; ``user_id`` is set automatically.
            **kwargs: Forwarded to both underlying ``save`` calls."""

        memory = self.get_or_create_user_memory(user_id)
        metadata = metadata or {}
        metadata["user_id"] = user_id

        item = memory.save(content=content, metadata=metadata, user_id=user_id, **kwargs)

        entity_mem = self.user_entities.get(user_id)
        if entity_mem:
            entity_mem.save(content=content, metadata=metadata, **kwargs)

        return item

    def search_user_memory(self, user_id: str, query: str, limit: int = 10, **kwargs) -> list:
        """Search the user's contextual memory only (not entities)."""

        memory = self.get_or_create_user_memory(user_id)
        return memory.search(query=query, limit=limit, **kwargs)

    def get_user_context(self, user_id: str) -> str:
        """Render a multi-section context block (preferences + summary + known entities)."""

        memory = self.get_or_create_user_memory(user_id)
        entity_mem = self.user_entities.get(user_id)

        context_parts = []

        prefs = self.get_user_preferences(user_id)
        if prefs:
            context_parts.append(f"User preferences: {prefs}")

        context_parts.append(memory.get_context_summary())

        if entity_mem and entity_mem.entities:
            entities = list(entity_mem.entities.keys())[:10]
            context_parts.append(f"Known entities: {', '.join(entities)}")

        return "\n\n".join(context_parts)

    def update_user_preferences(self, user_id: str, preferences: dict[str, Any]) -> None:
        """Merge ``preferences`` into the user's stored preferences dict and persist."""

        if user_id not in self.user_preferences:
            self.user_preferences[user_id] = self._get_default_preferences()

        self.user_preferences[user_id].update(preferences)
        self._save_preferences()

    def get_user_preferences(self, user_id: str) -> dict[str, Any]:
        """Return the user's preferences, falling back to the default template."""

        return self.user_preferences.get(user_id, self._get_default_preferences())

    def get_user_statistics(self, user_id: str) -> dict[str, Any]:
        """Return a statistics dict (memory counts, entity counts, preferences)."""

        stats = {
            "user_id": user_id,
            "total_memories": 0,
            "entities_known": 0,
            "preferences": self.get_user_preferences(user_id),
        }

        if user_id in self.user_memories:
            memory = self.user_memories[user_id]
            stats["total_memories"] = len(memory.short_term) + len(memory.long_term)
            stats["short_term_memories"] = len(memory.short_term)
            stats["long_term_memories"] = len(memory.long_term)

        if user_id in self.user_entities:
            entity_mem = self.user_entities[user_id]
            stats["entities_known"] = len(entity_mem.entities)
            stats["relationships"] = sum(len(rels) for rels in entity_mem.relationships.values())

        return stats

    def clear_user_memory(self, user_id: str) -> None:
        """Drop every tier and the preferences row for ``user_id``."""

        if user_id in self.user_memories:
            self.user_memories[user_id].clear()
            del self.user_memories[user_id]

        if user_id in self.user_entities:
            self.user_entities[user_id].clear()
            del self.user_entities[user_id]

        if user_id in self.user_preferences:
            del self.user_preferences[user_id]
            self._save_preferences()

    def _get_default_preferences(self) -> dict[str, Any]:
        """Return the canonical preference template for a freshly-seen user."""

        return {
            "response_style": "balanced",
            "verbosity": "normal",
            "technical_level": "intermediate",
            "language": "en",
            "timezone": "UTC",
            "memory_enabled": True,
            "max_context_items": 10,
        }

    def _save_preferences(self) -> None:
        """Persist the in-memory ``user_preferences`` map to storage."""

        if self.storage:
            self.storage.save("_user_preferences", self.user_preferences)

File: xerxes/src/memory/userMemory.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { createHash } from 'node:crypto'

import { scanContextContent } from '../security/promptScanner.js'
import {
  ContextualMemory,
  type ContextualSaveOptions,
} from './contextualMemory.js'
import { EntityMemory, type EntitySaveOptions } from './entityMemory.js'
import type { MemoryItem, MemoryMetadata, MemorySearchOptions } from './base.js'
import { NamespacedStorage, type MemoryStorage } from './storage.js'

const LEGACY_USER_PREFERENCES_KEY = '_user_preferences'
const USER_PREFERENCES_KEY_SUFFIX = 'user_preferences'

export type UserPreferences = Record<string, unknown>

export interface UserMemorySaveOptions extends ContextualSaveOptions, EntitySaveOptions {}

export interface UserMemoryStatistics {
  readonly entitiesKnown: number
  readonly longTermMemories?: number
  readonly preferences: UserPreferences
  readonly relationships?: number
  readonly shortTermMemories?: number
  readonly totalMemories: number
  readonly userId: string
}

/**
 * User-scoped facade that joins contextual recall, entity recall, and simple
 * persisted preferences without sharing an in-memory conversation between users.
 */
export class UserMemory {
  readonly userEntities = new Map<string, EntityMemory>()
  readonly userMemories = new Map<string, ContextualMemory>()
  readonly userPreferences = new Map<string, UserPreferences>()

  constructor(readonly storage?: MemoryStorage) {
    this.loadUsers()
  }

  clearUserMemory(userId: string): void {
    const memory = this.userMemories.get(userId)
    if (memory) {
      memory.clear()
      this.userMemories.delete(userId)
    }

    const entityMemory = this.userEntities.get(userId)
    if (entityMemory) {
      entityMemory.clear()
      this.userEntities.delete(userId)
    }

    if (this.userPreferences.delete(userId)) {
      // The per-user preference record goes with the user, and any
      // not-yet-migrated copy inside the legacy wholesale blob is scrubbed
      // too: otherwise a restart after clearing an inactive user would
      // resurrect stale preferences straight out of that blob.
      this.storage?.delete(preferenceStorageKey(userId))
      this.scrubLegacyPreferences(userId)
    }
  }

  /** Return a user-local contextual store, creating its entity and preference state if needed. */
  getOrCreateUserMemory(userId: string): ContextualMemory {
    const existing = this.userMemories.get(userId)
    if (existing) return existing

    // Persist under a per-user key namespace so tiers over a shared backend
    // only hydrate, search, and clear their own user's records. The namespace
    // segment is a hash of the user id, never the raw id: with raw ids the
    // prefix `user_a_` also matches user `a_b`'s keys, letting one user's
    // clear()/listKeys() delete or expose another user's records. The raw id
    // stays on the records themselves (metadata.user_id / options.userId).
    const storage = this.storage ? new NamespacedStorage(this.storage, userNamespace(userId)) : undefined
    const memory = new ContextualMemory({
      ...(storage ? { longTermStorage: storage } : {}),
      // Low-importance memories land in the short-term tier; without a durable
      // backend here they would vanish across restarts while only the
      // long-term side persisted.
      ...(storage ? { shortTermStorage: storage } : {}),
    })
    const entityMemory = new EntityMemory({
      ...(storage ? { storage } : {}),
    })
    this.userMemories.set(userId, memory)
    this.userEntities.set(userId, entityMemory)
    if (!this.userPreferences.has(userId)) this.userPreferences.set(userId, defaultPreferences())
    this.saveUserPreferences(userId)
    return memory
  }

  getUserContext(userId: string): string {
    const memory = this.getOrCreateUserMemory(userId)
    const entityMemory = this.userEntities.get(userId)
    const parts: string[] = []
    const preferences = this.getUserPreferences(userId)
    // Preference values and entity names are user- or tool-influenced data
    // recalled into prompts; neutralise embedded hostile instructions.
    if (Object.keys(preferences).length > 0) {
      parts.push(`User preferences: ${scanContextContent(JSON.stringify(preferences), `user preferences for ${userId}`)}`)
    }
    parts.push(memory.getContextSummary())
    if (entityMemory && Object.keys(entityMemory.entities).length > 0) {
      const names = Object.keys(entityMemory.entities).slice(0, 10).join(', ')
      parts.push(`Known entities: ${scanContextContent(names, `known entities for ${userId}`)}`)
    }
    return parts.join('\n\n')
  }

  getUserPreferences(userId: string): UserPreferences {
    return { ...(this.userPreferences.get(userId) ?? defaultPreferences()) }
  }

  getUserStatistics(userId: string): UserMemoryStatistics {
    const memory = this.userMemories.get(userId)
    const entityMemory = this.userEntities.get(userId)
    const statistics: UserMemoryStatistics = {
      userId,
      totalMemories: 0,
      entitiesKnown: 0,
      preferences: this.getUserPreferences(userId),
    }
    if (memory) {
      const shortTermMemories = memory.shortTerm.size
      const longTermMemories = memory.longTerm.size
      return {
        ...statistics,
        totalMemories: shortTermMemories + longTermMemories,
        shortTermMemories,
        longTermMemories,
        ...(entityMemory ? {
          entitiesKnown: Object.keys(entityMemory.entities).length,
          relationships: relationshipCount(entityMemory),
        } : {}),
      }
    }
    if (!entityMemory) return statistics
    return {
      ...statistics,
      entitiesKnown: Object.keys(entityMemory.entities).length,
      relationships: relationshipCount(entityMemory),
    }
  }

  saveMemory(
    userId: string,
    content: string,
    metadata: MemoryMetadata = {},
    options: UserMemorySaveOptions = {},
  ): MemoryItem {
    const memory = this.getOrCreateUserMemory(userId)
    const scopedMetadata = { ...metadata, user_id: userId }
    const scopedOptions: UserMemorySaveOptions = { ...options, userId }
    const item = memory.save(content, scopedMetadata, scopedOptions)
    this.userEntities.get(userId)?.save(content, scopedMetadata, scopedOptions)
    return item
  }

  searchUserMemory(userId: string, query: string, limit = 10, options: MemorySearchOptions = {}): MemoryItem[] {
    return this.getOrCreateUserMemory(userId).search(query, limit, undefined, options)
  }

  updateUserPreferences(userId: string, preferences: Readonly<UserPreferences>): void {
    this.userPreferences.set(userId, {
      ...(this.userPreferences.get(userId) ?? defaultPreferences()),
      ...preferences,
    })
    this.saveUserPreferences(userId)
  }

  private loadUsers(): void {
    if (!this.storage) return
    // Per-user namespaced records are authoritative once written.
    for (const key of this.storage.listKeys(USER_PREFERENCES_KEY_SUFFIX)) {
      if (!key.endsWith(`_${USER_PREFERENCES_KEY_SUFFIX}`)) continue
      const stored = this.storage.load(key)
      if (!isPreferenceRecord(stored)) continue
      this.userPreferences.set(stored.user_id, { ...stored.preferences })
    }
    // Legacy wholesale blob, migrated on read: distribute every entry to its
    // namespaced key once, then rewrite the blob without the entries whose
    // namespaced records now carry them (including copies that already did).
    // A blob left intact would resurrect users this facade cleared — its
    // stale copy outliving their deletion — so after migration it only ever
    // holds entries migration could not distribute. Never deleted outright:
    // older readers keep a valid, if aging, view.
    if (!this.storage.exists(LEGACY_USER_PREFERENCES_KEY)) return
    const stored = this.storage.load(LEGACY_USER_PREFERENCES_KEY)
    if (!isRecord(stored)) return
    let migratedAny = false
    const remaining: Record<string, unknown> = {}
    for (const [userId, preferences] of Object.entries(stored)) {
      const key = preferenceStorageKey(userId)
      if (isRecord(preferences) && !this.storage.exists(key)) {
        this.userPreferences.set(userId, { ...preferences })
        this.storage.save(key, preferenceRecord(userId, preferences))
        migratedAny = true
        continue
      }
      // Namespaced record already authoritative, or entry too malformed to
      // distribute; malformed shapes stay in the blob rather than being
      // silently discarded.
      if (isRecord(preferences)) migratedAny = true
      else remaining[userId] = preferences
    }
    if (migratedAny) this.storage.save(LEGACY_USER_PREFERENCES_KEY, remaining)
  }

  /** Remove one user's entry from the legacy wholesale blob, when present. */
  private scrubLegacyPreferences(userId: string): void {
    if (!this.storage?.exists(LEGACY_USER_PREFERENCES_KEY)) return
    const stored = this.storage.load(LEGACY_USER_PREFERENCES_KEY)
    if (!isRecord(stored) || !(userId in stored)) return
    const remaining = { ...stored }
    delete remaining[userId]
    this.storage.save(LEGACY_USER_PREFERENCES_KEY, remaining)
  }

  /**
   * Persist one user's preferences under their own namespaced key.
   *
   * A single shared blob made every writer's save a whole-map last-write-wins
   * across processes; distinct keys mean concurrent facades updating
   * different users never touch each other's records, and the read-merge
   * keeps same-user writers from dropping fields they never saw.
   */
  private saveUserPreferences(userId: string): void {
    if (!this.storage) return
    const stored = this.storage.load(preferenceStorageKey(userId))
    const merged: UserPreferences = {
      ...(isPreferenceRecord(stored) ? stored.preferences : {}),
      ...(this.userPreferences.get(userId) ?? {}),
    }
    this.storage.save(preferenceStorageKey(userId), preferenceRecord(userId, merged))
  }
}

function defaultPreferences(): UserPreferences {
  return {
    response_style: 'balanced',
    verbosity: 'normal',
    technical_level: 'intermediate',
    language: 'en',
    timezone: 'UTC',
    memory_enabled: true,
    max_context_items: 10,
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

/**
 * Collision-proof per-user storage namespace. Hashing the user id removes
 * prefix ambiguity between ids such as `a` and `a_b` and keeps raw ids out
 * of backend keys.
 */
function userNamespace(userId: string): string {
  const digest = createHash('sha256').update(userId, 'utf8').digest('hex').slice(0, 16)
  return `user_${digest}_`
}

/** Per-user preference record key: the namespaced prefix plus a fixed suffix. */
function preferenceStorageKey(userId: string): string {
  return `${userNamespace(userId)}${USER_PREFERENCES_KEY_SUFFIX}`
}

function preferenceRecord(userId: string, preferences: Readonly<UserPreferences>): { readonly preferences: UserPreferences; readonly user_id: string } {
  return { preferences: { ...preferences }, user_id: userId }
}

function isPreferenceRecord(value: unknown): value is { readonly preferences: UserPreferences; readonly user_id: string } {
  return isRecord(value)
    && typeof value.user_id === 'string'
    && value.user_id.length > 0
    && isRecord(value.preferences)
}

function relationshipCount(memory: EntityMemory): number {
  return Object.values(memory.relationships).reduce((count, pairs) => count + pairs.length, 0)
}
